# Configuration file format 4

# **************************************
# ******* EXPERIMENT INFORMATION *******
# **************************************

# Name of the experiment 
experience_name: minidancers_default

# ID for this particular run
run_id: 1

# Seed for the randomness of the experiment
seed: 10

# Length of the simulation in seconds
simulation_length: 500 # s

# **************************************
# ******** SCENARIO DEFINITION *********
# **************************************

# The number of robots in the fleet
robots_number: 3

# Position around which the robots will spawn
start_position:
  x: 0.0
  y: 0.0
  z: 10.0

# Parameter to configure the format of the exchanged messages between the 
cosimulation_mode: 'exchange_neighbors' # [exchange_neighbors, exchange_commands]

# If true, the module that invokes the controller will also publish the structs of the agents, for data recording 
publish_agent_structs: true

# **************************************
# ************* OBSTACLES **************
# **************************************

# Buildings are defined with their center (x, y), and their size in the three dimensions : (size_x, size_y, height)
auto_buildings: false # if false, "buildings" is used
distribution: uniform # [uniform, normal]
numb_obst: 23
radius_obst: 12
radius_stdev_obst: 1
height_obst: 30
height_stdev_obst: 15
arena_radius: 150
arena_center_x: 0
arena_center_y: 160

buildings:
- id: 1
  height: 30
  size_x: 15
  size_y: 30
  x: 20
  y: 40




# **************************************
# ********** ns-3 PARAMETERS  **********
# **************************************

ns3_print_simulation_advancement: false

# Wifi parameters
wifi_standard: '802.11ax' # [802.11g, 802.11n, 802.11ax]
wifi_type: 'SpectrumWifiPhy' # [YansWifiPhy, SpectrumWifiPhy]
propagation_loss_model: 'HybridBuildingsPropagationLossModel' # [LogDistancePropagationLossModel, HybridBuildingsPropagationLossModel], used only with YansWifiPhy
error_model_type: 'ns3::NistErrorRateModel'
phy_mode: 'HeMcs0' # 802.11g: [ErpOfdmRate54Mbps] 802.11n: [HtMcs0, HtMcs1, HtMcs2, HtMcs3, ..., HtMcs7] 802.11ax: [HeMcs0, HeMcs1, HeMcs2, HeMcs3, ..., HeMcs11]
short_guard_interval_supported: false
frequency: 5.2e9

# Routing parameters 
routing_algorithm: 'OLSR' # [AODV, DSDV, OLSR, BATMAN, DSR]
# Parameters for the "pseudo-routing" scheme where the routing in ns-3 is replaced with simple graph-traversal algorithms
pseudo_routing_algorithm: 'shortest_dist' # [shortest_dist, capacity_bottleneck, minimize_error_rate, predefined_chain] 
# If false, the route between source and sink, computed by the pseudo_routing_algorithm will be used to actually route the data. 
# If true, the pseudo_routing_algorithm still run, but the routing tables of the agents are handled by the routing_algorithm
use_real_routing_algorithm: false 

# Pseudo-routing algorithm specific parameters
shortest_dist:
  distance_threshold: 40 # m
  constant_before_threshold: 10

minimize_error_rate:
  broadcast_window_size: 0.5  # s

hybrid_dist_error_rate:
  distance_threshold: 60 # m
  constant_before_threshold: 10
  error_rate_exponent: 2
  distance_gain: 1
  error_rate_gain: 200

# Network applications
mission_flow:
  enable: true
  source_robot_ids:
    - 1
    - 2
  sink_robot_id: 0
  port: 7777
  packet_size: 1024     # bytes
  interval: 1000000      # us
  timeout: 2000000       # us
  start_time: 3.0       # s
  stop_time: 2000.0     # s
  flow_id: 1
  save_packets: true

broadcast_flow:
  enable: true
  port: 4000
  packet_size: 0       # bytes [The total size of the packet is packet_size + size(flocking_header). The flocking header is 49 bytes (role(1) + position(24) + velocity(24))]
  interval: 100000      # us
  timeout: 500000       # us
  start_time: 0.0       # s
  stop_time: 2000.0     # s
  flow_id: 2

# **************************************
# ******* CONTROLLER PARAMETERS  *******
# **************************************

# *** FLOCKING ***

# Indicates if the ROS flocking_controller should publish on a topic the command its sends.
controller_publish_cmd: true

# **** Constraints ****
expected_deconnexion_distance: 130.0

# **** Parameters ****
VAT_undefined_flocking_parameters:
  v_flock: 0.5
  v_max: 3
  v_sec_max: 3
  r_0_sec: 4.0
  v_sec_max_path: 3
  r_0_sec_path: 4.0
  a_frict: 4.16
  p_frict: 3.2
  r_0_frict: 85.3
  C_frict: 0.8
  v_frict: 0.63
  p_att: 0.08  # old 0.08
  r_0_att: 5
  p_rep: 0.13
  r_0_rep: 5
  a_shill: 53
  p_shill: 3.55
  r_0_shill: 1.0 # old 0.3
  v_shill: 13.622
  auto_propulsion_flag: false
  secondary_objective_flag: true
  # If use_deconnexion_distance_instead_of_p_att is true, p_att will be overriden so that the 
  # attraction speed can counter the v_sec_max at the expected_deconnexion_distance.
  use_deconnexion_distance_instead_of_p_att: false 
  use_squared_attraction_term: false
  p_los: 0.0
  r_los_obst_inflation: 0.0

VAT_mission_flocking_parameters:
  v_flock: 1.0
  v_max: 15
  v_sec_max: 6.0
  r_0_sec: 4.0
  v_sec_max_path: 6.0
  r_0_sec_path: 4.0
  a_frict: 4.16
  p_frict: 3.2
  r_0_frict: 85.3 # old 85.3
  C_frict: 0.8
  v_frict: 0.63
  p_att: 0.16  # old 0.08
  r_0_att: 30
  p_rep: 0.13
  r_0_rep: 20
  a_shill: 53
  p_shill: 3.55
  r_0_shill: 1.0 # old 0.3
  v_shill: 13.622
  auto_propulsion_flag: false
  secondary_objective_flag: true
  use_deconnexion_distance_instead_of_p_att: true
  use_squared_attraction_term: false
  p_los: 1.0
  r_los_obst_inflation: 5.0

VAT_potential_flocking_parameters:
  v_flock: 0.5
  v_max: 3
  v_sec_max: 3
  r_0_sec: 4.0
  v_sec_max_path: 3
  r_0_sec_path: 4.0
  a_frict: 4.16
  p_frict: 3.2
  r_0_frict: 85.3
  C_frict: 5.0
  v_frict: 0.63
  p_att: 0.16  # old 0.08
  r_0_att: 30
  p_rep: 0.13
  r_0_rep: 20
  a_shill: 53
  p_shill: 3.55
  r_0_shill: 1.0 # old 0.3
  v_shill: 13.622
  auto_propulsion_flag: false
  secondary_objective_flag: true
  use_deconnexion_distance_instead_of_p_att: false
  use_squared_attraction_term: false
  p_los: 0.0
  r_los_obst_inflation: 0.0

VAT_idle_flocking_parameters:
  v_flock: 0.5
  v_max: 10
  v_sec_max: 4
  r_0_sec: 4.0
  v_sec_max_path: 3
  r_0_sec_path: 4.0
  a_frict: 4.16
  p_frict: 3.2
  r_0_frict: 85.3
  C_frict: 0.8
  v_frict: 0.63
  p_att: 0.08
  r_0_att: 20
  p_rep: 0.13
  r_0_rep: 20
  a_shill: 53
  p_shill: 3.55
  r_0_shill: 1.0 # old 0.3
  v_shill: 13.622
  auto_propulsion_flag: false
  secondary_objective_flag: true
  use_deconnexion_distance_instead_of_p_att: false
  use_squared_attraction_term: false
  p_los: 0.0
  r_los_obst_inflation: 0.0

# secondary objectives of the drones, with the form robot_ID: [x, y, z, target(false)|sink(true)] 
# Be careful, there is no check that on the robot_ID in this list, if add a secondary objective for robot 10 and robot 10 does not exist, it will be ignored
auto_secondary_objectives: false # If true, the targets are automatically generated by the "generate_instances.py" script
num_secondary_objectives: 0
secondary_objectives: 
  - id: 1
    position: 
      x: -50
      y: 100
      z: 10
    is_sink: false
    assigned_agents: [1]
  - id: 0
    position: 
      x: 0
      y: 0
      z: 10
    is_sink: true
    assigned_agents: [0]

# target_altitude: 10.0 # Flocking is done at fixed altitude for now. If not defined, the flocking will be done in 3D.
min_altitude: 5.0 # Optional minimal altitude that, if set, will add an upward force if the UAV drops below the minimal altitude. Above this threshold, the UAV moves in 3D.

neighbor_timeout_value: 2000000 # Value after which the robot will be considered as lost (microseconds)

max_neighbors: 10 # Maximum number of neighbors considered by each robot (take the best ones)


# ******* LEADER PATH PLANNER *******
# When true, the secondary_objectives will be reached trough a A* start path planning
# When false, the agent will be attract by a direct force towards the goal.
use_planner: false

# Distance at which the planner will stop looking for closer grid cells.
goal_radius_tolerance: 1.0

# If the distance between the agent and its path is greater than distance_to_path_tolerance, the agent will replan a path.
distance_to_path_tolerance: 3.0

# Distance ahead of the closest point on the path that agent will be attract to. Similar to the carrot on a stick.
lookup_ahead_pursuit_distance: 5.0

# **** Occupancy grid for the planner ****

# Heigh of the grid in the global frame. 
grid_altitude: 10.0

# The size of the map is based on the smallest x and y position to the 
# largest x and y positions from obstalces and secondary objectives.
# The map inflation is a padding in meter to add around those dimensions.
map_inflation: 5.0

# Extra size to add on each size of obstacles for safe navigation. 
obstacle_inflation: 2.0

# Resolution in meter/cell of the occupancy grid.
map_resolution: 1.0

# ******* PRE-PLANNED FLIGHT *******
planned_trajectory:
  name: 'CIRCULAR'
  p1: 10.0
  p2: 0.1

  # name: 'INFINITY_SYMBOL'
  # p1: 10.0
  # p2: 0.1


# **************************************
# ************ STEINER TREE ************
# **************************************

stp_max_range: 40.0
stp_grid_resolution: 5.0
stp_grid_inflation: 10.0
stp_grid_altitude: 10.0
stp_obstacle_inflation: 5.0
stp_cost_function: 'piecewise_constant_exponential'
stp_solver_path: '/home/theotime/gits/scip-jack'
stp_solver_name: 'stp.linux.x86_64.gnu.opt.spx2'


# **************************************
# *************** GAZEBO ***************
# **************************************

world_sdf_path: src/physics_simulators/Gazebo/worlds/default.sdf       # (from root of ROS2 workspace)
update_rate: 20000 # in Hz | The objective rate at which Gazebo tries to run simulation loops. 

# Period of the command loop of the robots (Gazebo only)
# cmd_loop_period: 100000 # us

# The model of the robot. Must match the robots names in Gazebo in the following way : {robots_model}_{i} where i = 0..{robots_number}
# Example : x500_0, x500_1, x500_2, etc.
robots_model: x500

# **************************************
# ************ MINIDANCERS *************
# **************************************

publish_network_markers: false

use_localization_noise: false
localization_noise_stddev: 1.0

communication_range: 100.0 # m (only used when MiniDancers is launch in non-cosim mode)

mini_dancers_print_advancement: false


# For basic collisions
save_collisions_number: true
K_penetration_force: 200.0
K_restitution_force: 200.0
K_agent_penetration_force: 120.0
K_agent_restitution_force: 120.0
agent_radius: 0.5

# **************************************
# ************** ROBOTSIM **************
# **************************************

# Special variable used for Robotsim only (UDS socket connection between viragh_connector and Robotsim)
viragh_uds_server_address: /tmp/viragh_server_socket

# **************************************
# ********* DANCERS PARAMETERS *********
# **************************************

# Simulation timings. BE CAREFUL : sync_window must be divisible by phy_step_size, phy_step_size must be divisible by net_step_size 
# and phy_step_size must a multiple of step_length
sync_window: 10000 # in us | the duration between two position exchange and clock synchronization between physics and network simulators
phy_step_size: 10000 # in us | the duration of 1 physics simulator simulation loop
net_step_size: 10000 # in us

save_compute_time: false
enable_stats: false

net_use_uds: true
# network simulator socket: UDS
net_uds_server_address: /tmp/net_server_socket
# network simulator socket: TCP
net_ip_server_address: '127.0.0.1'
net_ip_server_port: 10000

phy_use_uds: true
# physics simulator socket: UDS
phy_uds_server_address: /tmp/phy_server_socket
# physics simulator socket: TCP
phy_ip_server_address: '127.0.0.1'
phy_ip_server_port: 10000